pkg/storagev2/inmemstore/inmembatch.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package inmemstore
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+)
+
+type batchOp struct {
+	delete bool
+	item   storage.Item
+}
+
+type Batch struct {
```

Comment:
We could but I think it's more valuable to keep them separate and let them evolve along their own logical lines, which might diverge in the future. A little copy is better than a little dependency, after all.

Reply:
There is problem when code starts do diverge. Imagine have any two (or more) similar functionalities which could benefit from same abstraction. Somewhere down do road, developer may find issue in one and fix it only in for this case, leaving other(s) vulnerable to same bug. 

Why would dependency to same batching mechanism be a problem?

Reply:
I wasn't able to figure out the commonalities of the two, besides, at this point we might want to leave this as it is now and optimize it later once we have the bandwidth.

Reply:
All what batch does is it collects operations to be preformed in `Commit`. Commit operation could have different implementations for `inmem` and `lvldb` stores. Pretty much everything just like in `inmemstore.Batch` currently does, except that `Commit` needs to support different implementations.

Reply:
The leveldb batch is writing the operations to `*leveldb.Batch` and the inmem one is writing to inmem map. The inmem batch is just a reference implementation. It doesnt provide the same level of functionality as `*leveldb.Batch`. The `WriteBatch` function is atomic.

Batch will be in place of a Store when user is using it. So it needs to provide the same API as the write operations. Just the additional Commit function. Different implementations can have different staging mechanisms. So the batch interface is flexible enough right now.

Reply:
All I am saying is that `inmemstore.Batch` and `leveldbstore.Batch` could use same mechanism because they are essentially just collecting what operations to execute. Only `Commit` step is different, which can be delegated to different implementations.

Reply:
I think this is one of those cases where it's easier to show than to tell.
Anyway, as Alok was saying - the inmem implementation is for reference, so there's no real and tangible benefit from optimizing things around it, at least right now.

Reply:
https://github.com/ethersphere/bee/pull/3314/files

Reply:
So I don't think the approach in this PR is better. Typically in a Batch, you would batch operations. So the implementation of how to batch the operations will be specific to the implementation. So the staging of the ops is also implementation dependent.

This is modeled after the levelDB.Batch interface. This also provides the same set of API (Put, Delete, Commit). This is also similar to what bolt and badger (and other KV stores) provide. The goal of the Batch interface here is to provide a consistent interface on top of these so that we can use any backend. Also, it needs to provide the Write APIs of the Store, so it can be used in place of that.

Also, these need to synchronize with the Store implementations as well. For eg. in the inmem implementation, we need access to the main mutex in order to push all the updates together. But this is not required for levelDB. But also in levelDB, we can't have multiple batches at a time, so we would need some sort of synchronization. There will be more such dependencies.

Reply:
This was draft PR just to show that single batch implementation would be sufficient for any specific implementation. I am not sure what are exact cons for this implementation? Because, synchronization shouldn't be a problem.

Either way, it's your call.